### PR TITLE
Introducing temporary errors. Refactored reestablish connection code …

### PR DESCRIPTION
**Pull request by rdelval**
…to use NewClien. Added reestablish connection test to end to end tests.

First PR on the way to fixing nil buffer error caused by Thrift library for 0.10.0.

-----------------------------------------
## Please read instructions below ##

Before submitting, please make sure you run a vagrant box running Aurora with the latest version shown in .auroraversion and run go test from the project root.

To run an Aurora Vagrant image, follow the instructions here:
http://aurora.apache.org/documentation/latest/getting-started/vagrant/

* Have you run goformat on the project before submitting? Yes

* Have you run go test on the project before submitting? Do all tests pass? Yes, Yes

* Does the Pull Request require a test to be added to the end to end tests? If so, has it been added? Yes. Yes

### CODE REVIEW

File: errors.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	temp, ok := err.(timeout)
+	return ok && temp.Timeout()
+}
+
+type TimeoutErr struct {
+	error
+	timeout bool
+}
+
+func (t TimeoutErr) Timeout() bool {
```
Comment (kkrishna) [inline errors.go]: why not use pointer?

Comment (rdelval) [reply errors.go]: Immutability, ideally no one should mess with the values inside the error after it's created.

File: retry.go [hunk 2]
```diff
@@ -59,24 +58,38 @@ func ExponentialBackoff(backoff Backoff, condition ConditionFunc) error {
...
 }
 
 // CheckAndRetryConn function takes realis client and a trhift API function to call and returns response and error
 // If Error from the APi call is Retry able . THe functions re establishes the connection with aurora by getting the latest aurora master from zookeeper.
 // If Error is retyable return resp and RetryConnErr error.
 func CheckAndRetryConn(r Realis, auroraCall AuroraThriftCall) (*aurora.Response, error) {
 	resp, cliErr := auroraCall()
-	if cliErr != nil /*&& (strings.Contains(cliErr.Error(), ConnRefusedErr) || strings.Contains(cliErr.Error(), NoLeaderFoundErr))*/ {
+
+	// TODO: Rerturn different error type based on the error that was returned by the API call
```
Comment (kkrishna) [inline retry.go]: typo..

Comment (rdelval) [reply retry.go]: Good catch

File: errors.go (hunk 1 above)
Comment (kkrishna) [reply errors.go]: interface has no setters and no public variable .. should suffice..

**Pull request merged**